### PR TITLE
enhancement(antithesis): Expand SDK assertions, hammer sketch

### PR DESCRIPTION
**Pull request by blt**
## Summary

This commit expands the SDK asserts in the SUT, allowing us to flag more
invariant flaws in runs. Of particular interest this commit introduces a
new driver, the 'sketchburst'. The goal of this driver is to transmit
sketch load in a way that 'bursts' load that will fall into a small
number of ddsketch bins.

This commit is also a demonstration of what antithesis sdk assertions
without some manner of shim or other way of tidying them up looks like
in a codebase, per recent review discussions.

## Change Type
- [ ] Bug fix
- [ ] New feature
- [x] Non-functional (chore, refactoring, docs)
- [ ] Performance

## How did you test this PR?

Confirmed function via antithesis shot.

## References

N/A

### CODE REVIEW

**Comment by blt** [comment]
* **#1799** <a href="https://app.graphite.com/github/pr/DataDog/saluki/1799?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 2  dependent PRs ([#1800](https://github.com/DataDog/saluki/pull/1800) <a href="https://app.graphite.com/github/pr/DataDog/saluki/1800?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#1826](https://github.com/DataDog/saluki/pull/1826) <a href="https://app.graphite.com/github/pr/DataDog/saluki/1826?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#1905** <a href="https://app.graphite.com/github/pr/DataDog/saluki/1905?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1785** <a href="https://app.graphite.com/github/pr/DataDog/saluki/1785?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/DataDog/saluki/1785?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#1782** <a href="https://app.graphite.com/github/pr/DataDog/saluki/1782?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1779** <a href="https://app.graphite.com/github/pr/DataDog/saluki/1779?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1768** <a href="https://app.graphite.com/github/pr/DataDog/saluki/1768?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 2 other dependent PRs ([#1769](https://github.com/DataDog/saluki/pull/1769) <a href="https://app.graphite.com/github/pr/DataDog/saluki/1769?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#1787](https://github.com/DataDog/saluki/pull/1787) <a href="https://app.graphite.com/github/pr/DataDog/saluki/1787?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#1767** <a href="https://app.graphite.com/github/pr/DataDog/saluki/1767?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1774** <a href="https://app.graphite.com/github/pr/DataDog/saluki/1774?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: bin/agent-data-plane/src/main.rs [hunk 1]
```diff
@@ -50,6 +50,30 @@ async fn main() -> Result<(), GenericError> {
...
+                .map(|s| (*s).to_string())
+                .or_else(|| payload.downcast_ref::<String>().cloned())
+                .unwrap_or_else(|| "<non-string panic payload>".to_string());
+            antithesis_sdk::assert_unreachable!(
+                "agent-data-plane panicked",
+                &serde_json::json!({ "message": message, "location": location })
+            );
+            default_hook(info);
+        }));
+    }
```
Comment (tobz) [inline bin/agent-data-plane/src/main.rs]: Could we actually stick this all in a standalone function called `initialize_antithesis` and gate that function call (and function definition) instead?

Just a lot of code/comments shoved at the very top of `main`.

File: bin/agent-data-plane/src/main.rs [hunk 2]
```diff
@@ -61,14 +85,26 @@ async fn main() -> Result<(), GenericError> {
...
-        .from_environment(PlatformSettings::get_env_var_prefix())
-        .error_context("Environment variable prefix should not be empty.")?
-        .bootstrap_generic();
+        .error_context("Failed to load Datadog Agent configuration file during bootstrap.")
+        .and_then(|loader| {
+            loader
+                .add_providers([DatadogRemapper::new()])
+                .from_environment(PlatformSettings::get_env_var_prefix())
+                .error_context("Environment variable prefix should not be empty.")
+        });
```
Comment (tobz) [inline bin/agent-data-plane/src/main.rs]: Could we maybe just pack this away in a standalone function called `load_bootstrap_config` -- everything that was here before -- which takes the bootstrap config path as a parameter.. and that way we can just do something like:

```rust
let maybe_bootstrap_config = load_bootstrap_config(&bootstrap_config_path);

// antithesis assertion here

let bootstrap_config = maybe_bootstrap_config?;
```

**Review by tobz — CHANGES_REQUESTED** [review]

File: bin/agent-data-plane/src/main.rs (hunk 1 above)
Comment (blt) [reply bin/agent-data-plane/src/main.rs]: Solid, I like that idea.

File: bin/agent-data-plane/src/main.rs (hunk 2 above)
Comment (blt) [reply bin/agent-data-plane/src/main.rs]: I'm into it.

File: lib/ddsketch/src/agent/sketch.rs [hunk 3]
```diff
@@ -711,6 +718,18 @@ fn trim_left(bins: &mut SmallVec<[Bin; 4]>, bin_limit: u16) {
...
     bins.drain(0..num_to_remove);
+
+    // This is the one place every mutating method routes through, so asserting here guards the bin-count bound for
+    // all of them.
+    #[cfg(feature = "antithesis")]
+    {
+        antithesis_sdk::assert_reachable!("DDSketch bin collapse reached");
+        antithesis_sdk::assert_always_less_than_or_equal_to!(
+            bins.len(),
+            bin_limit,
```
[the remark this answers is not in the document]
Comment (blt) [reply lib/ddsketch/src/agent/sketch.rs:730]: Already handled: `trim_left` casts `bin_limit` to `usize` at the top of the function, so both operands are `usize` by the time they reach the assertion. The `--features antithesis` build compiles cleanly. Resolving.

File: lib/saluki-context/src/resolver.rs [hunk 4]
```diff
@@ -727,6 +735,14 @@ impl TagsResolver {
...
             .or_else(|| {
                 self.allow_heap_allocations.then(|| {
+                    // Heap spill: with `allow_context_heap_allocations` true (the default), a full interner silently
+                    // falls back to the heap, so the bounded-memory guarantee no longer holds. Anchor that this path is
+                    // reached so the unbounded-growth behavior is observable.
+                    #[cfg(feature = "antithesis")]
+                    antithesis_sdk::assert_sometimes!(
+                        true,
+                        "context string interner spilled to the heap (unbounded under default config)"
+                    );
```
[the remark this answers is not in the document]
Comment (blt) [reply lib/saluki-context/src/resolver.rs:745]: Disambiguated: the `TagsResolver::intern` site now asserts "tag string interner spilled to the heap…", while the `ContextResolver::intern` site keeps "context string interner…". The two spill paths now carry distinct signatures, so counterexamples name which interner overflowed.

File: test/antithesis/harness/src/payload/dogstatsd/common.rs [hunk 5]
```diff
@@ -129,12 +129,19 @@ pub(crate) fn extend_choice(buf: &mut Vec<u8>, vibe: Vibe, compliant: &[&[u8]],
     }
 }
 
+/// Repeated-element counts (segments, tags): small body, feral `0`/large tail.
+const ELEMENT_COUNTS: &[u8] = &[1, 1, 2, 2, 3, 3, 4, 5, 6, 0, 127, 255];
+
+fn sample_count<R: Rng + ?Sized>(rng: &mut R) -> u8 {
+    ELEMENT_COUNTS[rng.random_range(0..ELEMENT_COUNTS.len())]
+}
```
[the remark this answers is not in the document]
Comment (blt) [reply test/antithesis/harness/src/payload/dogstatsd/common.rs]: Gated to feral: `ELEMENT_COUNTS` is split into `ELEMENT_COUNTS_CLEAN` (small body, no boundary cases) and `ELEMENT_COUNTS_FERAL` (adds the `0`/large tail). `sample_count` now selects the pool by vibe, so clean payloads no longer draw the boundary tail.

File: test/antithesis/harness/src/payload/dogstatsd.rs [hunk 6]
```diff
@@ -1,64 +1,76 @@
...
+//! <TAGS>        := #<TAG>(,<TAG>)*
+//! <MESSAGE>     := m:[^|\n]+
+//! ```
+//!
+//! # Name combinatorics
+//!
+//! A clean name is `c` segments from `COMPLIANT_WORD` (10 words) joined by
+//! `NAME_SEPARATORS` (4). Distinct names at count `c`: `10^c · 4^(c-1)`. `c` is
+//! sampled from `ELEMENT_COUNTS` — small body, with `0` and large boundary
+//! counts as the feral tail.
```
[the remark this answers is not in the document]
Comment (blt) [reply test/antithesis/harness/src/payload/dogstatsd.rs]: Updated: the module docs and probability table now describe the tail as feral-only — clean draws from `ELEMENT_COUNTS_CLEAN`, feral from `ELEMENT_COUNTS_FERAL`.

**Review by tobz — APPROVED** [review]

**Comment by blt** [comment]
### Merge activity

* **Jun 22, 9:29 PM UTC**: @blt merged this pull request with [Graphite](https://app.graphite.com/github/pr/DataDog/saluki/1785).
